pyomo/contrib/cp/scheduling_expr/step_function_expressions.py

```diff
@@ -0,0 +1,282 @@
+            return CumulativeFunction(_self.args + _other.args)
+    else:
+        raise TypeError("Cannot add object of class %s to object of class "
+                        "%s" % (_other.__class__, _self.__class__))
+
+def _generate_difference_expression(_self, _other):
+    # We check both because we call this function for the reverse operation as
+    # well.
+    if isinstance(_other, StepFunction) and isinstance(_self, StepFunction):
+        if _self.nargs() == len(_self._args_):
+            _self._args_.extend([NegatedStepFunction(a) for a in _other.args])
+            return CumulativeFunction(_self._args_, nargs=len(_self._args_))
+        else:
+            # we have to clone the list of _args_
+            return CumulativeFunction(_self.args + [NegatedStepFunction(a) for a
+                                                    in _other.args])
+    else:
+        raise TypeError("Cannot subtract object of class %s from object of "
+                        "class %s" % (_other.__class__, _self.__class__))
+
+class StepFunction(PyomoObject):
+    """
+    The base class for the step function expression system.
+    """
+    __slots__ = ()
+
+    def __add__(self, other):
+        return _generate_sum_expression(self, other)
+
+    def __radd__(self, other):
+        # Mathematically this doesn't make a whole lot of sense, but we'll call
+        # 0 a function and be happy so that sum() works as expected.
+        if other == 0:
+            return self
+        return _generate_sum_expression(other, self)
+
+    def __iadd__(self, other):
+        return _generate_sum_expression(self, other)
+
+    def __sub__(self, other):
+        return _generate_difference_expression(self, other)
+
+    def __rsub__(self, other):
+        return _generate_difference_expression(other, self)
+
+    def __isub__(self, other):
+        return _generate_difference_expression(self, other)
+
+    def within(self, bounds, times):
+        return AlwaysIn(self, bounds, times)
+
+    def nargs(self):
+        raise NotImplementedError(
+            f"Derived expression ({self.__class__}) failed to "
+            "implement nargs()")
+
+    @property
+    def args(self):
+        return self._args_[:self.nargs()]
+
+    # TODO: do we need clone?
+
+class Pulse(StepFunction):
+    """
+    A step function specified by an IntervalVar and an integer height that
+    has value 0 before the IntervalVar's start_time and after the
+    IntervalVar's end time and that takes the value specified by the 'height'
+    during the IntervalVar. (These are often used to model resource
+    constraints.)
+
+    Args:
+        interval_var (IntervalVar): the interval variable over which the
+            pulse function is non-zero
+        height (int): The value of the pulse function during the time
+            interval_var is scheduled
+    """
+
+    __slots__ = ('_args_')
+    def __init__(self, interval_var, height):
+        self._args_ = [self, interval_var, height]
```

Comment:
Because I was being waaay too much of a cheater and having these pretend to be `CumulativeFunction`s when I was adding and subtracting. I fixed that...